### PR TITLE
Granular timeouts

### PR DESCRIPTION
**Pull request by daniel-connectedenergy**
Lots of timing corrections and speed optimisations.

The main issue addresed were timeouts on TCP receive, occuring due to the configured timeouts expressed as seconds, which would inccur catastrophic cancelation on integer division by 1000 and having the receive interrupt bailing out early even though there was still lots of time left to receive.

Fixing this led to an overhaul of all the wait states used in the library and improvements in many routines (interrupt handler, GPRS setup, autobaud setup, FTP page reads, etc).

All these fixes make the library much more robust and faster - it has been tested on a CloudSolar device working properly at up to 115kbps; its GSM loop finishes in about 15 seconds after network registration (almost 18 seconds after system startup in best case) and an MQTT OTA update message can be processed and flashed in around 50 seconds since system startup. Test images for these are here: https://drive.google.com/drive/u/0/folders/1x9BJGyX0uNrS_BwMD0gFPxxmteaw86eR . There will be a related CloudSolar PR incoming.

Somewhat related: Next steps for improvement (which might still get caugt into this PR) would be removal of MBED code, as the tree has been tagged and we get a pointer where MBED code still lives in the history of the repo (and e.g CloudSolar MBED code uses hardcoded commit hashes anyway and knows what version of the library it uses). This will add a bit of sanity into the header file, which is terribly hard to follow at the moment.

### CODE REVIEW

**Review by JordanSilverman — COMMENTED** [review]
Lots of big changes here. I'm pretty concerned about reliability so would be keen to do a lot of testing on all of this.

My main question is around the choice for some of the timeouts. Looking at the SIM800 AT command documentation, it often mentions that you may need to wait a maximum of x seconds but we don't seem to be waiting for that long. One in particular is the AT+SAPBR command which has a max wait time of 85 seconds but the wait is cut down to 3.5 seconds. Why is that the case? If we carry on with our lives whenever you get a response, what is the harm in waiting for the suggested time? Maybe I'm reading it wrong though.

**Comment by JordanSilverman** [comment]
Testing with v0.4 device: 
1. I see it often gets stuck at the `MQTT Connect` stage before eventually getting unstuck. This seems to happen every now and then. I've got good signal strength in the office (-51) so that shouldn't be the issue. I've attached the log from the device from staging. Ignore the `TS MQTT resp: :AF Unsupported Content-Format` that was just because the filename for OTA was too long.
[cs_firmware_test.txt](https://github.com/lastmileICT/SIM800-AT/files/9665469/cs_firmware_test.txt)

2. After sending an OTA, the device was in prod mode and in bank 2. At this point, downlink messages weren't being picked up properly. It took 3 or 4 GSM loops for the load switching command from the server before it got processed properly and it didn't respond to an OTA command. I tried flashing staging firmware to bank 2 and it sometimes said 'no downlink message' when there definitely was a downlink message. Although it does eventually accept the messages.

3. After OTA-ing back to bank 1, in prod, device again sent loads of settings reports in a loop (seen in datadog) and is still not picking up downlink messages. After a reboot, it's now struggling to send any uplink messages let alone downlinks. I think this is similar behaviour in that occurred in the test device in India with the previous release.

4. Flashing staging back into bank 1 and everything works fine again, picking up downlink messages, perfectly sending uplinks.

It's difficult because I don't see consistency in the behaviour. Every time I flash or OTA it, it does something different but in general it seems pretty flaky I'm afraid. And for some reason, it's pretty much only getting into trouble in prod build which I really don't understand. I'm not really sure where to go from here or why/how there can be a difference between prod and staging.

**Comment by daniel-connectedenergy** [comment]
> Lots of big changes here. I'm pretty concerned about reliability so would be keen to do a lot of testing on all of this.
> 
> My main question is around the choice for some of the timeouts. Looking at the SIM800 AT command documentation, it often mentions that you may need to wait a maximum of x seconds but we don't seem to be waiting for that long. One in particular is the AT+SAPBR command which has a max wait time of 85 seconds but the wait is cut down to 3.5 seconds. Why is that the case? If we carry on with our lives whenever you get a response, what is the harm in waiting for the suggested time? Maybe I'm reading it wrong though.

If we wait for GPRS network registration, which we do - then bearer setup is never going to take more than ~200ms, so the documentation is dead wrong on this. I should have deleted the bogus comment. The 3.5 seconds timeout is more than enough, and if it fails, there are going to be 10 retries for a total of 35 seconds. If that still results in failure, the modem would get reset.

The TCP  connection by the way is done over the PDP context and not over the bearer. The bearer is only used internally by the modem for HTTP/FTP commands, so basically just for OTA.

**Comment by mcqtom** [comment]
I have pushed updates to this branch (and bumped the version in the CS repo) with some changes to fix two of the problems identified by @JordanSilverman:

- We will no longer get stuck in the MQTT connect phase as we do not try to connect until TCP is fully established.
- We should no longer miss downlink messages. I have added some dynamic delay to ensure that the TCP data has completely arrived before we try to read it. This has fixed what was a repeatable problem on my system.

The images at [the drive link](https://drive.google.com/drive/u/0/folders/1x9BJGyX0uNrS_BwMD0gFPxxmteaw86eR) have also been updated to my commit.

I have not been able to replicate the other issues but will try a little harder at that today.

**Comment by JordanSilverman** [comment]
Awesome, thanks Tom. Makes total sense to add those checks for increasing incoming TCP data. Just one question that I left inline.

I'll test it out on Monday but looks like it will work.

**Comment by mcqtom** [comment]
@daniel-connectedenergy I have now solved the issue with Jordan seeing multiple settings reports in a loop. The issue came down to the wait in send_tcp_data() being too short. Previously it was an arbitrary 1s. Now it is a dynamic up to 3s wait to see the two possible ack messages. I was getting SEND OK in about 500-600ms but Jordan was relatively repeatedly getting 1000-1200ms

I also changed one of the two #ifdefs related to quicksend as there was actually two different ones used in the function, CONFIG_SIM800_TCP_QUICKSEND and CONFIG_SIM800_USE_QUICKSEND. The former is actually present in Kconfig so I removed the latter.

Removing the arbitrary wait caused a knock on problem with downlink messages. Because the sending of the sub packet executed quickly and we got to process_message() faster than ever before, there was no data waiting in ip_rx_data() even with the new additions to check for data still arriving. I have introduced 1s wait here which I think is reasonable as it solves the problem on Jordans device and is only used once in the loop.

**Comment by JordanSilverman** [comment]
Really nice, seems to be a lot more stable even in dodgy network conditions at my house. Good job!

File: src/SIM800-AT-Zephyr.cpp [hunk 1]
```diff
@@ -689,13 +690,23 @@ int GPRS::send_tcp_data(const void *data, size_t len)
...
 #endif
 
-    prepare_for_rx(1000, NULL);
-    k_sleep(K_MSEC(1000));
+    // Sleep in 20ms slices until we get the ack back.
+    int wait_count = 3000 / 20;
+    while (wait_count--) {
+        k_sleep(K_MSEC(20));
+        if (ack_received){
+            LOG_DBG("Worked at %d", wait_count);
```
Comment (mcqtom) [inline src/SIM800-AT-Zephyr.cpp:703]: This DBG message should have been removed. Whoops

**Comment by daniel-connectedenergy** [comment]
Thanks for these contributions Tom, I am very happy with them and keen on having them in prod already :)

Let us please work a bit on the commit style, keeping the subject line to 72-80 characters max, one blank line and optional commit body if more info is required. Here is a random link describing the style I'm acustomed to: https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53 or thse guys: https://docs.zephyrproject.org/3.1.0/contribute/guidelines.html#commit-guidelines , but at least in this library we don't have too many subsystems, so keep the subject line with starting with a capital.  Basically your second sentence in both titles should become commit body. 

You can do this with, and it's about time you tried to use `git rebase --interactive`.

Also, maybe remove all the commented out debugging - leaving them in uncommented would pollute the console output and leaving them in commented are a bit of an eye sore.

This last bit can also be acomplished through an interactive rebase, so please ask for help if you need it.

File: src/SIM800-AT-Zephyr.cpp [hunk 2]
```diff
@@ -42,17 +42,29 @@ LOG_MODULE_REGISTER(sim800, CONFIG_GSM_LOG_LEVEL);
 
 int GPRS::ip_rx_data(void)
 {
+    char cmd[18];
+    uint16_t tcp_packet_len = 0;
+    uint16_t tcp_avail = 65535;
+
+    k_sleep(K_MSEC(1000)); // wait to make sure data has arrived
```
Comment (daniel-connectedenergy) [inline src/SIM800-AT-Zephyr.cpp:49]: I realise that this should be a concern for the calling application. It's a fix for a slow mqtt server sending unsollicited data at connection time, which shouldn't concern this library. Maybe move it in the calling function, right before calling this one ? Not feeling too strong about this one, so your call.

Comment (mcqtom) [reply src/SIM800-AT-Zephyr.cpp:49]: As for your comment it’s a fair point, but any ip_rx call from any situation would have to wait for the data to actually arrive? I think the best solution is a rx_available function in the library to do the waiting but that sounds like added complexity in terms of making it robust we don’t need right now. So perhaps your suggestion is the better compromise. Will change it
